mirgecom/viscous.py

```diff
@@ -0,0 +1,399 @@
+        \tau,-\mathbf{J}_\alpha],
+
+    with fluid velocity ($\mathbf{v}$), viscous stress tensor
+    ($\mathbf{\tau}$), heat flux ($\mathbf{q}$), and diffusive flux
+    for each species ($\mathbf{J}_\alpha$).
+
+    .. note::
+
+        The fluxes are returned as a :class:`mirgecom.fluid.ConservedVars`
+        object with a *dim-vector* for each conservation equation. See
+        :class:`mirgecom.fluid.ConservedVars` for more information about
+        how the fluxes are represented.
+
+    Parameters
+    ----------
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        The discretization to use
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state
+    cv: :class:`~mirgecom.fluid.ConservedVars`
+        Fluid state
+    grad_cv: :class:`~mirgecom.fluid.ConservedVars`
+        Gradient of the fluid state
+    grad_t: numpy.ndarray
+        Gradient of the fluid temperature
+
+    Returns
+    -------
+    :class:`~mirgecom.fluid.ConservedVars` or float
+        The viscous transport flux vector if viscous transport properties
+        are provided, scalar zero otherwise.
+    """
+    transport = eos.transport_model()
+    if transport is None:
+        return 0
+
+    dim = cv.dim
+    viscous_mass_flux = 0 * cv.momentum
+
+    j = diffusive_flux(discr, eos, cv, grad_cv)
+    heat_flux_diffusive = diffusive_heat_flux(discr, eos, cv, j)
+
+    tau = viscous_stress_tensor(discr, eos, cv, grad_cv)
+    viscous_energy_flux = (
+        np.dot(tau, cv.velocity)
+        - conductive_heat_flux(discr, eos, cv, grad_t)
+        - heat_flux_diffusive
+    )
+
+    return make_conserved(dim,
+            mass=viscous_mass_flux,
+            energy=viscous_energy_flux,
+            momentum=tau, species_mass=-j)
+
+
+def viscous_facial_flux(discr, eos, cv_tpair, grad_cv_tpair, grad_t_tpair,
+                        local=False):
+    """Return the viscous flux across a face given the solution on both sides.
+
+    Parameters
+    ----------
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        The discretization to use
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state
+    cv_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of :class:`~mirgecom.fluid.ConservedVars` with the fluid solution
+        on the faces
+    grad_cv_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of :class:`~mirgecom.fluid.ConservedVars` with the gradient of the
+        fluid solution on the faces
+    grad_t_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of temperature gradient on the faces.
+    local: bool
+        Indicates whether to skip projection of fluxes to "all_faces" or not. If
+        set to *False* (the default), the returned fluxes are projected to
+        "all_faces".  If set to *True*, the returned fluxes are not projected to
+        "all_faces"; remaining instead on the boundary restriction.
```

Comment:
I'd kind of prefer if this flag wasn't here. First, `local` as a name doesn't suggest what it does (at least to me, but maybe that's subjective). Second, typing out the `project` call at the call site seems clearer.

Reply:
(#509)  I'll get to this quickly - just wanted to note it in an issue.